File: tests/algorithms/x/Clojure/data_structures/trie/trie.clj
(ns main (:refer-clojure :exclude [new_trie remove_key insert insert_many find delete print_words test_trie print_results]))

(require 'clojure.set)

(defn in [x coll]
  (cond (string? coll) (clojure.string/includes? coll x) (map? coll) (contains? coll x) (sequential? coll) (some (fn [e] (= e x)) coll) :else false))

(defn padStart [s w p]
  (loop [out (str s)] (if (< (count out) w) (recur (str p out)) out)))

(defn indexOf [s sub]
  (let [idx (clojure.string/index-of s sub)] (if (nil? idx) -1 idx)))

(defn split [s sep]
  (clojure.string/split s (re-pattern sep)))

(def nowSeed (atom (let [s (System/getenv "MOCHI_NOW_SEED")] (if (and s (not (= s ""))) (Integer/parseInt s) 0))))

(declare new_trie remove_key insert insert_many find delete print_words test_trie print_results)

(def ^:dynamic _delete_ch nil)

(def ^:dynamic _delete_child_idx nil)

(def ^:dynamic _delete_children nil)

(def ^:dynamic _delete_new_children nil)

(def ^:dynamic _delete_node nil)

(def ^:dynamic _delete_should_delete nil)

(def ^:dynamic delete_nodes nil)

(def ^:dynamic delete_trie nil)

(def ^:dynamic dfs_node nil)

(def ^:dynamic find_ch nil)

(def ^:dynamic find_children nil)

(def ^:dynamic find_curr nil)

(def ^:dynamic find_i nil)

(def ^:dynamic find_node nil)

(def ^:dynamic find_nodes nil)

(def ^:dynamic insert_ch nil)

(def ^:dynamic insert_child_idx nil)

(def ^:dynamic insert_children nil)

(def ^:dynamic insert_curr nil)

(def ^:dynamic insert_i nil)

(def ^:dynamic insert_new_children nil)

(def ^:dynamic insert_new_node nil)

(def ^:dynamic insert_node nil)

(def ^:dynamic insert_nodes nil)

(def ^:dynamic insert_trie nil)

(def ^:dynamic remove_key_out nil)

(def ^:dynamic test_trie_ok nil)

(def ^:dynamic test_trie_t nil)

(def ^:dynamic test_trie_t2 nil)

(def ^:dynamic test_trie_t3 nil)

(def ^:dynamic test_trie_t4 nil)

(def ^:dynamic test_trie_trie nil)

(def ^:dynamic test_trie_words nil)

(defn new_trie []
  (try (throw (ex-info "return" {:v {:nodes [{:children {} :is_leaf false}]}})) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn remove_key [remove_key_m remove_key_k]
  (binding [remove_key_out nil] (try (do (set! remove_key_out {}) (doseq [key remove_key_m] (when (not= key remove_key_k) (set! remove_key_out (assoc remove_key_out key (nth remove_key_m key))))) (throw (ex-info "return" {:v remove_key_out}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn insert [insert_trie_p insert_word]
  (binding [insert_ch nil insert_child_idx nil insert_children nil insert_curr nil insert_i nil insert_new_children nil insert_new_node nil insert_node nil insert_nodes nil insert_trie nil] (do (set! insert_trie insert_trie_p) (set! insert_nodes (:nodes insert_trie)) (set! insert_curr 0) (set! insert_i 0) (while (< insert_i (count insert_word)) (do (set! insert_ch (nth insert_word insert_i)) (set! insert_child_idx (- 1)) (set! insert_children (:children (nth insert_nodes insert_curr))) (if (in insert_ch insert_children) (set! insert_child_idx (nth insert_children insert_ch)) (do (set! insert_new_node {:children {} :is_leaf false}) (set! insert_nodes (conj insert_nodes insert_new_node)) (set! insert_child_idx (- (count insert_nodes) 1)) (set! insert_new_children insert_children) (set! insert_new_children (assoc insert_new_children insert_ch insert_child_idx)) (set! insert_node (nth insert_nodes insert_curr)) (set! insert_node (assoc insert_node :children insert_new_children)) (set! insert_nodes (assoc insert_nodes insert_curr insert_node)))) (set! insert_curr insert_child_idx) (set! insert_i (+ insert_i 1)))) (set! insert_node (nth insert_nodes insert_curr)) (set! insert_node (assoc insert_node :is_leaf true)) (set! insert_nodes (assoc insert_nodes insert_curr insert_node)) (set! insert_trie (assoc insert_trie :nodes insert_nodes)))))

(defn insert_many [insert_many_trie insert_many_words]
  (doseq [w insert_many_words] (insert insert_many_trie w)))

(defn find [find_trie find_word]
  (binding [find_ch nil find_children nil find_curr nil find_i nil find_node nil find_nodes nil] (try (do (set! find_nodes (:nodes find_trie)) (set! find_curr 0) (set! find_i 0) (while (< find_i (count find_word)) (do (set! find_ch (nth find_word find_i)) (set! find_children (:children (nth find_nodes find_curr))) (when (not (in find_ch find_children)) (throw (ex-info "return" {:v false}))) (set! find_curr (nth find_children find_ch)) (set! find_i (+ find_i 1)))) (set! find_node (nth find_nodes find_curr)) (throw (ex-info "return" {:v (:is_leaf find_node)}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn _delete [delete_trie delete_word _delete_idx _delete_pos]
  (binding [_delete_ch nil _delete_child_idx nil _delete_children nil _delete_new_children nil _delete_node nil _delete_should_delete nil] (try (do (when (= _delete_pos (count delete_word)) (do (set! _delete_node (nth delete_nodes _delete_idx)) (when (= (:is_leaf _delete_node) false) (throw (ex-info "return" {:v false}))) (set! _delete_node (assoc _delete_node :is_leaf false)) (alter-var-root (var delete_nodes) (fn [_] (assoc delete_nodes _delete_idx _delete_node))) (throw (ex-info "return" {:v (= (count (:children _delete_node)) 0)})))) (set! _delete_node (nth delete_nodes _delete_idx)) (set! _delete_children (:children _delete_node)) (set! _delete_ch (nth delete_word _delete_pos)) (when (not (in _delete_ch _delete_children)) (throw (ex-info "return" {:v false}))) (set! _delete_child_idx (nth _delete_children _delete_ch)) (set! _delete_should_delete (_delete _delete_child_idx (+ _delete_pos 1))) (set! _delete_node (nth delete_nodes _delete_idx)) (when _delete_should_delete (do (set! _delete_new_children (remove_key (:children _delete_node) _delete_ch)) (set! _delete_node (assoc _delete_node :children _delete_new_children)) (alter-var-root (var delete_nodes) (fn [_] (assoc delete_nodes _delete_idx _delete_node))) (throw (ex-info "return" {:v (and (= (count _delete_new_children) 0) (= (:is_leaf _delete_node) false))})))) (alter-var-root (var delete_nodes) (fn [_] (assoc delete_nodes _delete_idx _delete_node))) (throw (ex-info "return" {:v false}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn delete [delete_trie_p delete_word]
  (binding [delete_nodes nil delete_trie nil] (try (do (set! delete_trie delete_trie_p) (set! delete_nodes (:nodes delete_trie)) (_delete delete_trie delete_word 0 0) (set! delete_trie (assoc delete_trie :nodes delete_nodes))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn dfs [print_words_trie dfs_idx dfs_word]
  (binding [dfs_node nil] (do (set! dfs_node (get (:nodes print_words_trie) dfs_idx)) (when (:is_leaf dfs_node) (println dfs_word)) (doseq [key (keys (:children dfs_node))] (dfs (get (:children dfs_node) key) (str dfs_word key))))))

(defn print_words [print_words_trie]
  (dfs print_words_trie 0 ""))

(defn test_trie []
  (binding [test_trie_ok nil test_trie_t nil test_trie_t2 nil test_trie_t3 nil test_trie_t4 nil test_trie_trie nil test_trie_words nil] (try (do (set! test_trie_words ["banana" "bananas" "bandana" "band" "apple" "all" "beast"]) (set! test_trie_trie (new_trie)) (insert_many test_trie_trie test_trie_words) (set! test_trie_ok true) (doseq [w test_trie_words] (set! test_trie_ok (and test_trie_ok (find test_trie_trie w)))) (set! test_trie_ok (and test_trie_ok (find test_trie_trie "banana"))) (set! test_trie_t (find test_trie_trie "bandanas")) (set! test_trie_ok (and test_trie_ok (= test_trie_t false))) (set! test_trie_t2 (find test_trie_trie "apps")) (set! test_trie_ok (and test_trie_ok (= test_trie_t2 false))) (set! test_trie_ok (and test_trie_ok (find test_trie_trie "apple"))) (set! test_trie_ok (and test_trie_ok (find test_trie_trie "all"))) (delete test_trie_trie "all") (set! test_trie_t3 (find test_trie_trie "all")) (set! test_trie_ok (and test_trie_ok (= test_trie_t3 false))) (delete test_trie_trie "banana") (set! test_trie_t4 (find test_trie_trie "banana")) (set! test_trie_ok (and test_trie_ok (= test_trie_t4 false))) (set! test_trie_ok (and test_trie_ok (find test_trie_trie "bananas"))) (throw (ex-info "return" {:v test_trie_ok}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn print_results [print_results_msg print_results_passes]
  (if print_results_passes (println (str print_results_msg " works!")) (println (str print_results_msg " doesn't work :("))))

(def ^:dynamic main_trie (new_trie))

(defn -main []
  (let [rt (Runtime/getRuntime)
    start-mem (- (.totalMemory rt) (.freeMemory rt))
    start (System/nanoTime)]
      (print_results "Testing trie functionality" (test_trie))
      (System/gc)
      (let [end (System/nanoTime)
        end-mem (- (.totalMemory rt) (.freeMemory rt))
        duration-us (quot (- end start) 1000)
        memory-bytes (Math/abs ^long (- end-mem start-mem))]
        (println (str "{\n  \"duration_us\": " duration-us ",\n  \"memory_bytes\": " memory-bytes ",\n  \"name\": \"main\"\n}"))
      )
    ))

(-main)

File: tests/algorithms/x/Pascal/data_structures/trie/trie.pas
{$mode objfpc}
program Main;
uses SysUtils, Variants, fgl;
type Node = record
  children: specialize TFPGMap<string, integer>;
  is_leaf: boolean;
end;
type Trie = record
  nodes: array of Node;
end;
type NodeArray = array of Node;
type StrArray = array of string;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
procedure panic(msg: string);
begin
  writeln(msg);
  halt(1);
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  remove_key_key_idx: integer;
  _delete_node_var: Node;
  _delete_children: specialize TFPGMap<string, integer>;
  _delete_ch: string;
  _delete_child_idx: integer;
  _delete_child_idx_idx: integer;
  _delete_should_delete: Variant;
  _delete_new_children: specialize TFPGMap<string, integer>;
  dfs_node_var: Node;
  dfs_key: string;
  dfs_key_idx: integer;
  trie_var: Trie;
  word: string;
  passes: boolean;
  m: specialize TFPGMap<string, integer>;
  msg: string;
  k: string;
  words: StrArray;
function Map1(): specialize TFPGMap<string, integer>; forward;
function makeTrie(nodes: NodeArray): Trie; forward;
function makeNode(children: specialize TFPGMap<string, integer>; is_leaf: boolean): Node; forward;
function new_trie(): Trie; forward;
function remove_key(m: specialize TFPGMap<string, integer>; k: string): specialize TFPGMap<string, integer>; forward;
procedure insert(trie_var: Trie; word: string); forward;
procedure insert_many(trie_var: Trie; words: StrArray); forward;
function find(trie_var: Trie; word: string): boolean; forward;
procedure delete(trie_var: Trie; word: string); forward;
procedure print_words(trie_var: Trie); forward;
function test_trie(): boolean; forward;
procedure print_results(msg: string; passes: boolean); forward;
function Map1(): specialize TFPGMap<string, integer>;
begin
  Result := specialize TFPGMap<string, integer>.Create();
end;
function makeTrie(nodes: NodeArray): Trie;
begin
  Result.nodes := nodes;
end;
function makeNode(children: specialize TFPGMap<string, integer>; is_leaf: boolean): Node;
begin
  Result.children := children;
  Result.is_leaf := is_leaf;
end;
function new_trie(): Trie;
begin
  exit(makeTrie([makeNode(Map1(), false)]));
end;
function remove_key(m: specialize TFPGMap<string, integer>; k: string): specialize TFPGMap<string, integer>;
var
  remove_key_out: specialize TFPGMap<string, integer>;
  remove_key_key: string;
begin
  remove_key_out := specialize TFPGMap<string, integer>.Create();
  for remove_key_key_idx := 0 to (m.Count - 1) do begin
  remove_key_key := m.Keys[remove_key_key_idx];
  if remove_key_key <> k then begin
  remove_key_out[remove_key_key] := m[remove_key_key];
end;
end;
  exit(remove_key_out);
end;
procedure insert(trie_var: Trie; word: string);
var
  insert_nodes: array of Node;
  insert_curr: integer;
  insert_i: integer;
  insert_ch: string;
  insert_child_idx: integer;
  insert_children: specialize TFPGMap<string, integer>;
  insert_new_node: Node;
  insert_new_children: specialize TFPGMap<string, integer>;
  insert_node_var: Node;
begin
  insert_nodes := trie_var.nodes;
  insert_curr := 0;
  insert_i := 0;
  while insert_i < Length(word) do begin
  insert_ch := word[insert_i+1];
  insert_child_idx := -1;
  insert_children := insert_nodes[insert_curr].children;
  if insert_children.IndexOf(insert_ch) <> -1 then begin
  insert_child_idx := insert_children[insert_ch];
end else begin
  insert_new_node := makeNode(Map1(), false);
  insert_nodes := concat(insert_nodes, [insert_new_node]);
  insert_child_idx := Length(insert_nodes) - 1;
  insert_new_children := insert_children;
  insert_new_children[insert_ch] := insert_child_idx;
  insert_node_var := insert_nodes[insert_curr];
  insert_node_var.children := insert_new_children;
  insert_nodes[insert_curr] := insert_node_var;
end;
  insert_curr := insert_child_idx;
  insert_i := insert_i + 1;
end;
  insert_node_var := insert_nodes[insert_curr];
  insert_node_var.is_leaf := true;
  insert_nodes[insert_curr] := insert_node_var;
  trie_var.nodes := insert_nodes;
end;
procedure insert_many(trie_var: Trie; words: StrArray);
var
  insert_many_w: string;
begin
  for insert_many_w in words do begin
  insert(trie_var, insert_many_w);
end;
end;
function find(trie_var: Trie; word: string): boolean;
var
  find_nodes: array of Node;
  find_curr: integer;
  find_i: integer;
  find_ch: string;
  find_children: specialize TFPGMap<string, integer>;
  find_node_var: Node;
begin
  find_nodes := trie_var.nodes;
  find_curr := 0;
  find_i := 0;
  while find_i < Length(word) do begin
  find_ch := word[find_i+1];
  find_children := find_nodes[find_curr].children;
  if not find_children.IndexOf(find_ch) <> -1 then begin
  exit(false);
end;
  find_curr := find_children[find_ch];
  find_i := find_i + 1;
end;
  find_node_var := find_nodes[find_curr];
  exit(find_node_var.is_leaf);
end;
procedure delete(trie_var: Trie; word: string);
var
  delete_nodes: array of Node;
  delete_idx: integer;
  delete_pos: integer;
  function _delete(delete_idx: integer; delete_pos: integer): boolean;
begin
  if delete_pos = Length(word) then begin
  _delete_node_var := delete_nodes[delete_idx];
  if _delete_node_var.is_leaf = false then begin
  exit(false);
end;
  _delete_node_var.is_leaf := false;
  delete_nodes[delete_idx] := _delete_node_var;
  exit(Length(_delete_node_var.children) = 0);
end;
  _delete_node_var := delete_nodes[delete_idx];
  _delete_children := _delete_node_var.children;
  _delete_ch := word[delete_pos+1];
  if not _delete_children.IndexOf(_delete_ch) <> -1 then begin
  exit(false);
end;
  _delete_child_idx_idx := _delete_children.IndexOf(_delete_ch);
  if _delete_child_idx_idx <> -1 then begin
  _delete_child_idx := _delete_children.Data[_delete_child_idx_idx];
end else begin
  _delete_child_idx := 0;
end;
  _delete_should_delete := _delete(_delete_child_idx, delete_pos + 1);
  _delete_node_var := delete_nodes[delete_idx];
  if _delete_should_delete then begin
  _delete_new_children := remove_key(_delete_node_var.children, _delete_ch);
  _delete_node_var.children := _delete_new_children;
  delete_nodes[delete_idx] := _delete_node_var;
  exit((Length(_delete_new_children) = 0) and (_delete_node_var.is_leaf = false));
end;
  delete_nodes[delete_idx] := _delete_node_var;
  exit(false);
end;
begin
  delete_nodes := trie_var.nodes;
  _delete(0, 0);
  trie_var.nodes := delete_nodes;
end;
procedure print_words(trie_var: Trie);
var
  print_words_idx: integer;
  procedure dfs(print_words_idx: integer; word: string);
begin
  dfs_node_var := trie_var.nodes[print_words_idx];
  if dfs_node_var.is_leaf then begin
  writeln(word);
end;
  for dfs_key_idx := 0 to (.Count - 1) do begin
  dfs_key := .Keys[dfs_key_idx];
  dfs(dfs_node_var.children[dfs_key], word + dfs_key);
end;
end;
begin
  dfs(0, '');
end;
function test_trie(): boolean;
var
  test_trie_words: array of string;
  test_trie_trie_var: Trie;
  test_trie_ok: boolean;
  test_trie_w: string;
  test_trie_t: boolean;
  test_trie_t2: boolean;
  test_trie_t3: boolean;
  test_trie_t4: boolean;
begin
  test_trie_words := ['banana', 'bananas', 'bandana', 'band', 'apple', 'all', 'beast'];
  test_trie_trie_var := new_trie();
  insert_many(test_trie_trie_var, test_trie_words);
  test_trie_ok := true;
  for test_trie_w in test_trie_words do begin
  test_trie_ok := test_trie_ok and find(test_trie_trie_var, test_trie_w);
end;
  test_trie_ok := test_trie_ok and find(test_trie_trie_var, 'banana');
  test_trie_t := find(test_trie_trie_var, 'bandanas');
  test_trie_ok := test_trie_ok and (test_trie_t = false);
  test_trie_t2 := find(test_trie_trie_var, 'apps');
  test_trie_ok := test_trie_ok and (test_trie_t2 = false);
  test_trie_ok := test_trie_ok and find(test_trie_trie_var, 'apple');
  test_trie_ok := test_trie_ok and find(test_trie_trie_var, 'all');
  delete(test_trie_trie_var, 'all');
  test_trie_t3 := find(test_trie_trie_var, 'all');
  test_trie_ok := test_trie_ok and (test_trie_t3 = false);
  delete(test_trie_trie_var, 'banana');
  test_trie_t4 := find(test_trie_trie_var, 'banana');
  test_trie_ok := test_trie_ok and (test_trie_t4 = false);
  test_trie_ok := test_trie_ok and find(test_trie_trie_var, 'bananas');
  exit(test_trie_ok);
end;
procedure print_results(msg: string; passes: boolean);
begin
  if passes then begin
  writeln(msg + ' works!');
end else begin
  writeln(msg + ' doesn''t work :(');
end;
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  trie_var := new_trie();
  print_results('Testing trie functionality', test_trie());
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.
